File: apps/apple/TeslaSync/shared-surfaces/MaintenanceBanner.Model.swift
//
//  MaintenanceBanner.Model.swift
//  TeslaSync — P4 shared surface · 0127 · MaintenanceBanner (Apple)
//
//  The state-holder seam (P1/S8), the telemetry seam (P1/S11), the i18n facade (P1/S10), and the pure
//  projection for the maintenance / degraded-mode banner. The view binds through
//  `MaintenanceBannerModel`; no networking lives in the view. The web `MaintenanceBanner` wraps
//  `useSystemHealth()`, hides itself when `mode === 'ok'` (or when the current snapshot is dismissed),
//  and otherwise renders a sticky banner with a 1 Hz live countdown to `maintenance_until`. The native
//  model keeps the same contract: a source emits the coalesced health snapshot (resolved mode +
//  maintenance message / until / updated-at + the query's load / connectivity state), the model derives
//  the resolved projection over it (including the per-snapshot dismissal and its reset on a fresh
//  snapshot), drives the countdown off an injected clock, and auto-refreshes once when the feed
//  transitions to stale.
//

import Foundation
import Observation
import OSLog

// MARK: - Surface identity

/// The diagnostics surface slug, kept on a pure type so the model (and its isolated unit tests) need not
/// reference the SwiftUI surface view to emit `view.opened`.
public enum MaintenanceBannerSurface {
    public static let slug = "MaintenanceBanner"
}

// MARK: - Telemetry seam (P1/S11 diagnostics contract)

/// Emits the `view.opened` product-analytics event for the surface. The default logs via `os.Logger`;
/// the production app injects an adapter that forwards to the shared-core diagnostics sink (consent
/// gated + redacted there). The slug is a static, non-identifying constant.
public protocol MaintenanceBannerTelemetry: Sendable {
    func viewOpened(surface: String)
}

/// `os.Logger`-backed default that records the surface open as a redaction-safe `view.opened` event.
public struct OSLogMaintenanceBannerTelemetry: MaintenanceBannerTelemetry {
    private let logger: Logger

    public init(subsystem: String = "io.teslasync.app", category: String = "diagnostics") {
        logger = Logger(subsystem: subsystem, category: category)
    }

    public func viewOpened(surface: String) {
        logger.info("view.opened surface=\(surface, privacy: .public)")
    }
}

// MARK: - Connectivity (P4 leaf freshness axis)

/// The freshness of the bound health feed — the orthogonal connectivity axis rendered as the freshness
/// chip. `live` hides the chip; `stale` / `offline` show it.
public enum MaintenanceBannerConnection: String, Sendable, Equatable, CaseIterable {
    case live
    case stale
    case offline
}

// MARK: - Input snapshot (health state + feed lifecycle)

/// One coalesced snapshot of the surface's inputs — the web `useSystemHealth` result (`mode` +
/// `maintenance_message` / `maintenance_until` / `maintenance_updated_at`) plus the query's lifecycle
/// (`hasData` = the web `data != null`, an error message, and connectivity). `hasData == false` with no
/// error is the initial-load window (web `data` undefined); `hasData == true` with `mode == "ok"` is the
/// resolved no-banner state (web `return null`).
public struct MaintenanceBannerInput: Sendable, Equatable {
    /// Web `data?.mode` — the resolved backend service mode (`ok` / `degraded` / `maintenance`).
    public var mode: String
    /// Web `data?.maintenance_message` — the operator message folded into the body copy when non-blank.
    public var message: String
    /// Web `data?.maintenance_until` — the ISO-8601 window end driving the countdown.
    public var until: String
    /// Web `data?.maintenance_updated_at` — the operator update instant keying the dismissal.
    public var updatedAt: String
    /// Web `data != null` — a `/system/health` payload has resolved at least once.
    public var hasData: Bool
    /// The most recent feed error, surfaced only when no payload has resolved yet (web keeps last data).
    public var errorMessage: String?
    public var connection: MaintenanceBannerConnection

    public init(
        mode: String = "ok",
        message: String = "",
        until: String = "",
        updatedAt: String = "",
        hasData: Bool = false,
        errorMessage: String? = nil,
        connection: MaintenanceBannerConnection = .live
    ) {
        self.mode = mode
        self.message = message
        self.until = until
        self.updatedAt = updatedAt
        self.hasData = hasData
        self.errorMessage = errorMessage
        self.connection = connection
    }

    /// The resolved `MaintenanceBannerServiceMode` for the raw backend value (web `data?.mode ?? 'ok'`).
    public var serviceMode: MaintenanceBannerServiceMode {
        MaintenanceBannerServiceMode.forRaw(mode)
    }

    /// Web `!data || mode === 'ok'` inverted — a real banner-worthy snapshot has resolved.
    public var hasActiveBanner: Bool {
        hasData && serviceMode.isActive
    }

    /// The dismissal fingerprint for this snapshot (web `fingerprint(mode, message, until, updatedAt)`).
    public var fingerprint: String {
        MaintenanceBannerFingerprint.make(mode: mode, message: message, until: until, updatedAt: updatedAt)
    }
}

// MARK: - Resolved view-state (web render branches + P4 leaf contract)

/// The data payload for the `.banner` phase — the fully-derived banner: the mode (for the tone / icon /
/// `data-mode` parity), the precomposed title + body, the dismissal fingerprint carried through so the
/// view's dismiss action knows what to persist, the parsed window-end millis driving the countdown, and
/// the SF Symbol name. A pure value so the view is a function of it and snapshot tests assert it directly.
public struct MaintenanceBannerData: Sendable, Equatable {
    public let mode: MaintenanceBannerServiceMode
    public let isMaintenance: Bool
    public let title: String
    public let body: String
    public let fingerprint: String
    public let untilMs: Double?
    public let systemImageName: String

    public init(
        mode: MaintenanceBannerServiceMode,
        isMaintenance: Bool,
        title: String,
        body: String,
        fingerprint: String,
        untilMs: Double?,
        systemImageName: String
    ) {
        self.mode = mode
        self.isMaintenance = isMaintenance
        self.title = title
        self.body = body
        self.fingerprint = fingerprint
        self.untilMs = untilMs
        self.systemImageName = systemImageName
    }
}

/// The resolved, view-ready state — `phase` selects the body; for the banner phase the derived `data`
/// payload is pre-computed so the view is a pure function of this value (plus the time-dependent
/// countdown text the model maintains separately off the clock).
public struct MaintenanceBannerResolved: Sendable, Equatable {
    public enum Phase: Sendable, Equatable {
        case loading
        case empty
        case error(String)
        case banner
    }

    public let phase: Phase
    public let data: MaintenanceBannerData?

    public init(phase: Phase, data: MaintenanceBannerData?) {
        self.phase = phase
        self.data = data
    }
}

// MARK: - Projection (web render branches + P4 leaf contract)

/// Pure projection from the input snapshot + the dismissed fingerprint to the resolved view-state — the
/// native port of the web banner's control flow plus the P4 leaf contract:
///   • no payload yet → `error` if the initial read failed, else `loading` (the web pre-resolve window).
///   • a resolved payload always governs: a background feed error is ignored while data is present (web
///     TanStack keeps the last `data`), so an active banner stays on screen through a transient failure.
///   • `mode === 'ok'` → the friendly `empty` (web `return null`) — the native improvement over rendering
///     nothing, never a blank box.
///   • a dismissed snapshot (the fingerprint matches the dismissed key) → `empty` (web `return null`).
///   • an active, non-dismissed snapshot → the `banner` with its pre-composed copy.
/// Unit tested across every branch.
public enum MaintenanceBannerProjection {
    public static func resolve(
        input: MaintenanceBannerInput,
        dismissedKey: String?,
        strings: MaintenanceBannerResolve = MaintenanceBannerStrings.string
    ) -> MaintenanceBannerResolved {
        guard input.hasData else {
            if let message = input.errorMessage, !message.isEmpty {
                return MaintenanceBannerResolved(phase: .error(message), data: nil)
            }
            return MaintenanceBannerResolved(phase: .loading, data: nil)
        }
        guard input.hasActiveBanner else {
            return MaintenanceBannerResolved(phase: .empty, data: nil)
        }
        if dismissedKey == input.fingerprint {
            return MaintenanceBannerResolved(phase: .empty, data: nil)
        }
        return MaintenanceBannerResolved(phase: .banner, data: payload(input, strings))
    }

    private static func payload(
        _ input: MaintenanceBannerInput,
        _ strings: MaintenanceBannerResolve
    ) -> MaintenanceBannerData {
        let mode = input.serviceMode
        let isMaintenance = mode.isMaintenance
        return MaintenanceBannerData(
            mode: mode,
            isMaintenance: isMaintenance,
            title: MaintenanceBannerMessage.title(isMaintenance: isMaintenance, strings: strings),
            body: MaintenanceBannerMessage.body(isMaintenance: isMaintenance, message: input.message, strings: strings),
            fingerprint: input.fingerprint,
            untilMs: MaintenanceBannerInstant.parseMs(input.until),
            systemImageName: mode.systemImageName
        )
    }
}

// MARK: - State-holder (P1/S8 layer)

/// The surface's observable view-model. Subscribes to a `MaintenanceBannerSource`, recomputes the resolved
/// projection, exposes a render `phase` + the resolved view-state, the `connection` axis, and the live
/// `countdownText`. It mirrors the web component's lifecycle precisely: it resets a stale dismissal when
/// the upstream snapshot fingerprint changes (web `useEffect`), persists a dismissal per snapshot (web
/// `sessionStorage`), drives a 1 Hz countdown off the injected clock only while a banner with a window
/// end is showing (web `setInterval` gated on `mode !== 'ok' && untilMs !== null`), and auto-refreshes
/// once when the feed transitions to stale.
@MainActor
@Observable
public final class MaintenanceBannerModel {
    public private(set) var resolved = MaintenanceBannerResolved(phase: .loading, data: nil)
    public private(set) var connection: MaintenanceBannerConnection = .live
    public private(set) var countdownText: String?

    public var phase: MaintenanceBannerResolved.Phase {
        resolved.phase
    }

    public var data: MaintenanceBannerData? {
        resolved.data
    }

    @ObservationIgnored private let source: any MaintenanceBannerSource
    @ObservationIgnored private let telemetry: any MaintenanceBannerTelemetry
    @ObservationIgnored private let strings: MaintenanceBannerResolve
    @ObservationIgnored private let dismissalStore: any MaintenanceBannerDismissalStore
    @ObservationIgnored private let clock: any MaintenanceBannerClock
    @ObservationIgnored private var started = false
    @ObservationIgnored private var countdownRunning = false
    @ObservationIgnored private var dismissedKey: String?
    @ObservationIgnored private var lastInput = MaintenanceBannerInput()

    public init(
        source: any MaintenanceBannerSource,
        telemetry: any MaintenanceBannerTelemetry = OSLogMaintenanceBannerTelemetry(),
        strings: @escaping MaintenanceBannerResolve = MaintenanceBannerStrings.string,
        dismissalStore: any MaintenanceBannerDismissalStore = SessionMaintenanceBannerDismissalStore.shared,
        clock: any MaintenanceBannerClock = SystemMaintenanceBannerClock()
    ) {
        self.source = source
        self.telemetry = telemetry
        self.strings = strings
        self.dismissalStore = dismissalStore
        self.clock = clock
        dismissedKey = dismissalStore.read()
        source.onUpdate = { [weak self] input in self?.apply(input) }
    }

    /// Begins observing and emits the `view.opened` diagnostics event. Idempotent.
    public func start() {
        guard !started else { return }
        started = true
        telemetry.viewOpened(surface: MaintenanceBannerSurface.slug)
        source.start()
    }

    /// Stops observing the upstream feed and tears down the countdown tick.
    public func stop() {
        started = false
        clock.stop()
        countdownRunning = false
        source.stop()
    }

    /// Re-requests the upstream snapshot (freshness chip + post-window "refresh to confirm").
    public func refresh() {
        source.refresh()
    }

    /// Web dismiss button → persists the current snapshot's fingerprint and hides the banner. A later
    /// snapshot with a different fingerprint re-surfaces it (the dismissal reset in `apply`).
    public func dismiss() {
        guard resolved.phase == .banner, let fingerprint = resolved.data?.fingerprint else { return }
        dismissalStore.write(fingerprint)
        dismissedKey = fingerprint
        recompute(lastInput)
        updateCountdown()
    }

    private func apply(_ input: MaintenanceBannerInput) {
        lastInput = input
        if let key = dismissedKey, key != input.fingerprint {
            dismissedKey = nil
        }
        recompute(input)
        let previous = connection
        connection = input.connection
        if input.connection == .stale, previous != .stale {
            source.refresh()
        }
        updateCountdown()
    }

    private func recompute(_ input: MaintenanceBannerInput) {
        resolved = MaintenanceBannerProjection.resolve(input: input, dismissedKey: dismissedKey, strings: strings)
    }

    private func updateCountdown() {
        guard resolved.phase == .banner, let untilMs = resolved.data?.untilMs else {
            countdownText = nil
            if countdownRunning {
                clock.stop()
                countdownRunning = false
            }
            return
        }
        refreshCountdownText(untilMs: untilMs)
        if !countdownRunning {
            countdownRunning = true
            clock.start { [weak self] in self?.tick() }
        }
    }

    private func tick() {
        guard resolved.phase == .banner, let untilMs = resolved.data?.untilMs else {
            updateCountdown()
            return
        }
        refreshCountdownText(untilMs: untilMs)
    }

    private func refreshCountdownText(untilMs: Double) {
        let nowMs = clock.now().timeIntervalSince1970 * 1000
        countdownText = MaintenanceBannerMessage.countdown(remainingMs: untilMs - nowMs, strings: strings)
    }
}

// MARK: - Localization facade (P1/S10) — web `t(key, default)`

/// Resolves the surface's strings by key with the web English fallback, so the views hold no hardcoded
/// literals. Keys live in the "MaintenanceBanner" table, folded into the app `Localizable.xcstrings`
/// catalog at integration time; kept per-surface so each parallel prompt owns its own strings.
public enum MaintenanceBannerStrings {
    public static let table = "MaintenanceBanner"

    public static let string: MaintenanceBannerResolve = { key, fallback in
        NSLocalizedString(key, tableName: table, bundle: .main, value: fallback, comment: "")
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/MaintenanceBanner/MaintenanceBannerModel.kt
// Pure, framework-free model + projection + diagnostics for the MaintenanceBanner shared surface — the native
// analogue of every value the web component derives before returning JSX
// (web/src/components/feedback/MaintenanceBanner.tsx). No Compose, no Android framework, no HTTP: every
// declaration here is exercised off-device by the :android:testReleaseUnitTest gate, keeping the composable a
// thin render layer over these pure functions.
//
// WHAT THE WEB SOURCE IS (and therefore the COMPLETE branch set this surface reproduces). The web file is a
// sticky top-of-app banner that polls `/api/v1/system/health` (via `useSystemHealth`) and renders when the
// resolved service `mode` is `degraded` or `maintenance`. It carries a title, a body (the operator message or
// a per-mode default), a live one-second countdown to `maintenance_until`, and a per-snapshot dismiss control
// keyed on a fingerprint of `maintenance_updated_at` (or mode/message/until when updated_at is absent) so a
// dismissal sticks for THAT banner state but a freshly-pushed operator banner re-surfaces. When `mode === 'ok'`
// or the health read has not resolved (`!data`) the web returns `null` — the banner is simply absent.
//
// HOW THAT MAPS ONTO THE NATIVE WIRED STATE (P1/S8, ADR-002/013). `useSystemHealth` is the shared
// `AdminStore.systemHealth()` feed — the `/system/health` cache-then-network `Resource` every native Admin
// surface already shares (the same feed `SystemHealthWidget` binds). [MaintenanceBannerSnapshot.fromJson] is
// the cached-payload → typed-projection data adapter the source plugs in, and [MaintenanceBannerProjection]
// folds that snapshot together with the current clock and the per-snapshot dismissal into the exact branch the
// banner renders. The data-envelope states the web hides behind `!data` are reproduced honestly through the
// ADR-013 freshness contract: a cold load with nothing cached and a hard failure with nothing cached both keep
// the banner absent (web `!data`), while an active window served from a stale cache or after a failed refresh
// keeps the banner visible with an explicit "Stale" chip (offline / last-known) rather than presenting stale
// state as live.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/MaintenanceBanner — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen segment is illegal in a package identifier), so the package intentionally diverges
// from the path — exactly as the sibling shared surfaces do. `MatchingDeclarationName` is suppressed for the
// co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.maintenancebanner

import io.teslasync.shared.core.diagnostics.Logger
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import java.time.Instant

/**
 * Canonical registry metadata for the MaintenanceBanner surface. The diagnostics [SLUG] is emitted with the
 * one-shot `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`MaintenanceBanner`);
 * [ID] is the stable `viewModel` key the host binds the surface with.
 */
object MaintenanceBannerRegistration {
    /** Stable surface id (also the `viewModel` key the host binds the surface with). */
    const val ID: String = "maintenance-banner"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = "MaintenanceBanner"
}

/**
 * The resolved service-mode bucket the banner keys its variant off — the native port of the web
 * `data?.mode` values:
 *  - [Ok] — the banner is hidden (web `mode === 'ok'`, plus the absent / unknown read);
 *  - [Degraded] — the sky-toned "Service is degraded" variant;
 *  - [Maintenance] — the amber-toned "Scheduled maintenance" variant.
 */
enum class ServiceMode {
    Ok,
    Degraded,
    Maintenance,
    ;

    companion object {
        /** The `ok` wire value — banner hidden (web `mode === 'ok'`). */
        const val RAW_OK: String = "ok"

        /** The `degraded` wire value — the sky-toned variant. */
        const val RAW_DEGRADED: String = "degraded"

        /** The `maintenance` wire value — the amber-toned variant. */
        const val RAW_MAINTENANCE: String = "maintenance"

        /**
         * Maps the raw `/system/health` `mode` string onto the typed bucket; anything that is not an explicit
         * `degraded` / `maintenance` (including `ok`, `null`, and any unknown value) collapses to [Ok] — the
         * banner-hidden default (web `data?.mode ?? 'ok'`).
         */
        fun fromRaw(raw: String?): ServiceMode =
            when (raw) {
                RAW_MAINTENANCE -> Maintenance
                RAW_DEGRADED -> Degraded
                else -> Ok
            }
    }
}

/**
 * The slice of the `/system/health` document this banner reads — the native mirror of the four `SystemHealth`
 * fields the web component touches (web/src/types/admin.ts). The full DTO also carries `status`, `components`,
 * `databaseSize`, `tableCount`, and `source`, none of which the banner renders (those back the separate
 * SystemHealthWidget), so they are deliberately omitted (DRY — the model carries only what the surface renders).
 *
 * @property rawMode the raw `mode` string (`ok` default); kept verbatim because the dismissal [fingerprint]
 *   hashes it exactly as the web does, and [mode] resolves it to the typed bucket.
 * @property message the operator banner text (web `maintenance_message`); blank/absent falls back at render.
 * @property untilIso the ISO-8601 auto-clear instant (web `maintenance_until`), or `""` for none.
 * @property updatedAtIso the ISO-8601 snapshot stamp (web `maintenance_updated_at`); the primary dismissal key.
 * @property present whether the `/system/health` read resolved to an object at all — distinguishes a cold load
 *   / hard failure with nothing cached (web `!data` → banner absent) from a resolved `ok` snapshot.
 */
data class MaintenanceBannerSnapshot(
    val rawMode: String,
    val message: String,
    val untilIso: String,
    val updatedAtIso: String,
    val present: Boolean,
) {
    /** The resolved service-mode bucket (web `data?.mode ?? 'ok'`). */
    val mode: ServiceMode get() = ServiceMode.fromRaw(rawMode)

    /** Whether this snapshot drives a visible banner variant (web `mode !== 'ok'`), given the read resolved. */
    val isActive: Boolean get() = present && mode != ServiceMode.Ok

    companion object {
        private const val KEY_MODE = "mode"
        private const val KEY_MESSAGE = "maintenance_message"
        private const val KEY_UNTIL = "maintenance_until"
        private const val KEY_UPDATED_AT = "maintenance_updated_at"

        /** The pre-resolution snapshot: no read yet (web `!data`); the banner stays absent. */
        val ABSENT: MaintenanceBannerSnapshot =
            MaintenanceBannerSnapshot(rawMode = ServiceMode.RAW_OK, message = "", untilIso = "", updatedAtIso = "", present = false)

        /**
         * Parses the shared store's raw `/system/health` [json] into the typed slice this banner reads — the
         * data adapter the source plugs into the feed. Snake_case keys are read verbatim (the shared
         * `AdminRepository` carries the server JSON unchanged); a missing `mode` defaults to `ok` (banner
         * hidden), absent string fields collapse to `""`, and a non-object payload yields `null` so a
         * `Resource.Loading` with no cache stays a first-load (web `!data`) rather than a resolved snapshot.
         */
        fun fromJson(json: JsonElement?): MaintenanceBannerSnapshot? {
            val obj = json as? JsonObject ?: return null
            return MaintenanceBannerSnapshot(
                rawMode = obj.string(KEY_MODE) ?: ServiceMode.RAW_OK,
                message = obj.string(KEY_MESSAGE) ?: "",
                untilIso = obj.string(KEY_UNTIL) ?: "",
                updatedAtIso = obj.string(KEY_UPDATED_AT) ?: "",
                present = true,
            )
        }

        private fun JsonObject.string(key: String): String? = (this[key] as? JsonPrimitive)?.contentOrNull
    }
}

/**
 * The localized-at-the-boundary countdown line the banner shows under the body — the native mirror of the web
 * `countdown` ternary. Carries only the locale-stable formatted duration ([EndsIn.formatted]); the surrounding
 * "Ends in {time}" / "Ending now" / "Window has ended…" copy resolves through the P1/S10 catalog at render.
 */
sealed interface Countdown {
    /** A still-running window (web `remaining > 1000`): "Ends in {formatted}" where [formatted] is e.g. "29m 59s". */
    data class EndsIn(
        val formatted: String,
    ) : Countdown

    /** The imminent edge (web `remaining > -1000`): "Ending now". */
    data object EndingNow : Countdown

    /** The window's end instant is already in the past (web else branch): "Window has ended; refresh to confirm.". */
    data object Ended : Countdown
}

/**
 * The fully-resolved, render-ready view of the banner — the native analogue of everything the web component
 * computes before returning JSX. It folds the `/system/health` snapshot, the current clock, the dismissal, and
 * the feed-freshness envelope into the exact branch decisions the composable renders, so the surface stays a
 * thin render layer and each rule has one test-pinned definition.
 *
 * @property visible whether the banner renders at all (web `!data || mode === 'ok' ? null` plus the dismissal).
 * @property maintenance the amber maintenance variant (web `isMaintenance`); else the sky degraded variant.
 * @property message the operator banner text, or `null`/blank to fall back to the per-mode default at render.
 * @property countdown the resolved countdown line, or `null` when there is no parseable `maintenance_until`.
 * @property stale whether the rendered window is served stale / last-known (offline) — drives the "Stale" chip.
 * @property currentKey the dismissal fingerprint of this snapshot; the composable hands it back on dismiss.
 */
data class MaintenanceBannerRender(
    val visible: Boolean,
    val maintenance: Boolean,
    val message: String?,
    val countdown: Countdown?,
    val stale: Boolean,
    val currentKey: String,
) {
    /** The sky degraded variant (web `!isMaintenance`); the complement of [maintenance], for symmetric reads. */
    val degraded: Boolean get() = !maintenance

    /** Whether the freshness "Stale" chip should render — a visible, stale, last-known window (ADR-013). */
    val showStaleChip: Boolean get() = visible && stale
}

/**
 * Pure projection of a [MaintenanceBannerSnapshot] (plus the clock, dismissal, and freshness envelope) into the
 * [MaintenanceBannerRender] — the native mirror of every branch the web component takes. Framework-free so the
 * whole contract is covered by the JVM unit gate without a Compose host.
 */
object MaintenanceBannerProjection {
    /** Below this many ms remaining the line reads "Ending now" rather than a duration (web `remaining > 1000`). */
    private const val ENDS_IN_THRESHOLD_MS: Long = 1_000L

    /** Below this many ms remaining (i.e. >1s past the end) the line reads "ended" (web `remaining > -1000`). */
    private const val ENDING_NOW_FLOOR_MS: Long = -1_000L

    private const val MILLIS_PER_SECOND: Long = 1_000L
    private const val SECONDS_PER_MINUTE: Long = 60L
    private const val SECONDS_PER_HOUR: Long = 3_600L
    private const val PAD_WIDTH: Int = 2

    /**
     * The per-snapshot dismissal fingerprint — a 1:1 port of the web `fingerprint(mode, message, until,
     * updatedAt)`: a present `updated_at` is the authoritative key (`u:<stamp>`), otherwise a deterministic
     * hash of the mode / message / until content (`s:<mode>|<message>|<until>`). This keeps an operator's
     * "I just pushed a new banner" workflow honest — any change to the snapshot yields a new key, so a prior
     * dismissal does not swallow a fresh announcement.
     */
    fun fingerprint(snapshot: MaintenanceBannerSnapshot): String =
        if (snapshot.updatedAtIso.isNotEmpty()) {
            "u:${snapshot.updatedAtIso}"
        } else {
            "s:${snapshot.rawMode}|${snapshot.message}|${snapshot.untilIso}"
        }

    /**
     * Folds the [snapshot], the current [nowMs], the [dismissedKey], and the feed-freshness flags into the
     * render state. Mirrors the web visibility gate (`!data || mode === 'ok'` → hidden, plus the
     * `dismissedKey === currentKey` → hidden), the maintenance/degraded variant split, the message → per-mode
     * default fallback, and the countdown ternary. [stale] is the ADR-013 offline / last-known flag (the web
     * banner has no such chip; native discloses it rather than presenting stale state as live).
     */
    fun render(
        snapshot: MaintenanceBannerSnapshot,
        nowMs: Long,
        dismissedKey: String?,
        stale: Boolean = false,
    ): MaintenanceBannerRender {
        val key = fingerprint(snapshot)
        val dismissed = dismissedKey != null && dismissedKey == key
        val visible = snapshot.isActive && !dismissed
        return MaintenanceBannerRender(
            visible = visible,
            maintenance = snapshot.mode == ServiceMode.Maintenance,
            message = snapshot.message.trim().ifBlank { null },
            countdown = countdownFor(parseUntil(snapshot.untilIso), nowMs),
            stale = stale,
            currentKey = key,
        )
    }

    /**
     * Resolves the countdown line from the parsed end instant [untilMs] and the current [nowMs] — the native
     * port of the web `remaining` ternary: a future end → [Countdown.EndsIn] with the short duration, the
     * sub-second edge → [Countdown.EndingNow], a past end → [Countdown.Ended], and no end → `null`.
     */
    fun countdownFor(
        untilMs: Long?,
        nowMs: Long,
    ): Countdown? {
        if (untilMs == null) return null
        val remaining = untilMs - nowMs
        return when {
            remaining > ENDS_IN_THRESHOLD_MS -> Countdown.EndsIn(formatRemaining(remaining))
            remaining > ENDING_NOW_FLOOR_MS -> Countdown.EndingNow
            else -> Countdown.Ended
        }
    }

    /**
     * Renders a positive [ms] duration as the web's zero-padded short form: "Hh MMm" above an hour, "Mm SSs"
     * above a minute, else "Ss". Locale-stable (no grouping separators), so it lives in the pure model and the
     * surrounding localized template ("Ends in {time}") is applied at the render boundary.
     */
    fun formatRemaining(ms: Long): String {
        val total = (ms / MILLIS_PER_SECOND).coerceAtLeast(0L)
        val hours = total / SECONDS_PER_HOUR
        val minutes = (total % SECONDS_PER_HOUR) / SECONDS_PER_MINUTE
        val seconds = total % SECONDS_PER_MINUTE
        return when {
            hours > 0L -> "${hours}h ${minutes.pad()}m"
            minutes > 0L -> "${minutes}m ${seconds.pad()}s"
            else -> "${seconds}s"
        }
    }

    /**
     * Parses an ISO-8601 `maintenance_until` [raw] to epoch millis — the native port of the web
     * `Date.parse(maintenance_until)` with its `Number.isFinite` guard: a blank or unparseable value yields
     * `null` (the web `null` branch), so the countdown affordance is simply not shown.
     */
    fun parseUntil(raw: String): Long? =
        if (raw.isBlank()) {
            null
        } else {
            runCatching { Instant.parse(raw).toEpochMilli() }.getOrNull()
        }

    private fun Long.pad(): String = toString().padStart(PAD_WIDTH, '0')
}

/** The stable, dot-namespaced diagnostics event emitted once when the surface opens (P1/S11). */
const val EVENT_VIEW_OPENED: String = "view.opened"

/** The structured-field key carrying the surface slug on every diagnostic. */
const val FIELD_SURFACE: String = "surface"

/**
 * Emits the one PII-safe `view.opened` diagnostic carrying only the surface [MaintenanceBannerRegistration.SLUG]
 * (P1/S11) — never the maintenance message, mode, or end time — so a diagnostics line can never leak the
 * fleet's operational posture. Kept free of Compose so it is unit-tested with a recording [Logger]; the
 * ViewModel calls it once per surface open.
 */
fun recordMaintenanceBannerOpened(logger: Logger) {
    logger.info(EVENT_VIEW_OPENED, mapOf(FIELD_SURFACE to MaintenanceBannerRegistration.SLUG))
}
